### PR TITLE
Finding template placeholder

### PR DESCRIPTION
**Pull request by Siryu6**
**Description**

Small enhancement of Finding template allowing to keep "original" content of finding with a placeholder.
Matching based on Title of findings.

- Using `{{original}}` placeholder allows to keep original outputs of the finding and can be include inside the template.
  eg: Templating output of a tool automatically and keep including tool output in the finding.

**Test results**

Ran "Finding" related tests.

**Documentation**

N/A

**Checklist**

This checklist is for your information.

- [x] Make sure to rebase your PR against the very latest `dev`.
- [x] Features/Changes should be submitted against the `dev`.
- [ ] Bugfixes should be submitted against the `bugfix` branch.
- [x] Give a meaningful name to your PR, as it may end up being used in the release notes.
- [x] Your code is flake8 compliant.
- [x] Your code is python 3.11 compliant.
- [ ] If this is a new feature and not a bug fix, you've included the proper documentation in the docs at https://github.com/DefectDojo/django-DefectDojo/tree/dev/docs as part of this PR.
- [x] Model changes must include the necessary migrations in the dojo/db_migrations folder.
- [ ] Add applicable tests to the unit tests.
- [x] Add the proper label to categorize your PR.

### CODE REVIEW

File: dojo/db_migrations/0191_finding_template_findings_to_replace_and_more.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+        migrations.AddField(
+            model_name='finding_template',
+            name='findings_to_replace',
+            field=models.TextField(blank=True, help_text='Title of findings you want to replace with this template separate by semicolon.', max_length=1000, null=True),
+        ),
+        migrations.AddField(
+            model_name='finding_template',
+            name='list_replace',
+            field=models.BooleanField(default=False, help_text='Apply template to all findings listed in "Findings to replace". (Update will overwrite all filled fields. You can merge field using {original} placeholder.)'),
+        ),
```
Comment (kiblik) [inline dojo/db_migrations/0191_finding_template_findings_to_replace_and_more.py:22]: I suppose these lines could be replaced by lines from `0192` and the whole `0192` can be removed.

**Review by kiblik — CHANGES_REQUESTED** [review]

File: dojo/db_migrations/0191_finding_template_findings_to_replace_and_more.py (hunk 1 above)
Comment (damiencarol) [reply dojo/db_migrations/0191_finding_template_findings_to_replace_and_more.py:22]: I agree, this could be merged.

Comment (Siryu6) [reply dojo/db_migrations/0191_finding_template_findings_to_replace_and_more.py:22]: Hi ! 
Not sure to understand what you are talking about 😅.
If I can help on anything just tell me :)

Comment (kiblik) [reply dojo/db_migrations/0191_finding_template_findings_to_replace_and_more.py:22]: Like this https://github.com/Siryu6/django-DefectDojo/pull/1

**Comment by Siryu6** [comment]
Do you need any further info on that ?

**Review by kiblik — APPROVED** [review]

**Comment by mtesauro** [comment]
@Siryu6 There's a bunch of failing tests that need to be addressed before this can be reviewed further

**Comment by Siryu6** [comment]
Hi @mtesauro ! 
Thanks for the checks it appears that I missed that 😅

**Comment by kiblik** [comment]
Do you mind adding some unittest, please?

**Comment by Siryu6** [comment]
Not sure what I failed here as we just added some tests :')

File: dojo/finding/views.py [hunk 2]
```diff
@@ -2238,6 +2242,28 @@ def apply_cwe_mitigation(apply_to_findings, template, update=True):
                 impact=template.impact,
                 references=template.references,
             )
+        elif template.list_replace:
+            for entry in template.findings_to_replace.split(';;'):
+                to_update = Finding.objects.filter(
+                    active=True,
+                    title=entry,
```
Comment (kiblik) [inline dojo/finding/views.py:2249]: The original implementation used also 
```python
                verified=True,
                cwe=template.cwe,
```
why it is excluded here?

**Review by kiblik — CHANGES_REQUESTED** [review]

**Comment by kiblik** [comment]
Hi @Siryu6,

regarding 
> python: can't open file '/home/runner/work/django-DefectDojo/django-DefectDojo/.github/scripts/git_protect.py': [Errno 2] No such file or directory

it is not connected to your work, you can ignore it for now. Guys are trying to fix it by #8953 and #8954

Regarding
> Do you mind adding some unittest, please?

I see you added nice integration tests (dir `tests/`). But can you add unit tests (dir `unittests/`) as well?
You can test `template_replace_placeholder` with different values (empty fields, placeholder `{{original}}`, fields with `{{original}}` or `{{{{original}}}}`), how does it behave based on different content of `findings_to_replace` and each of the impacted fields (empty, one `;`, two `;`, three `;`, four `;`, )?

And do you mind describing this behaviour in documentation or field description as well? E.g. I see it impacts fields `severity`, `cvssv3`, `description`, `mitigation`, `impact` and `references` but it is not obvious until you do not check the code.

Thank you

**Comment by kiblik** [comment]
> Hi @kiblik
> 
> Thanks for your answer ! I'll do some doc about that !
> 
> Quick question, would you mind check that unittest (https://github.com/Siryu6/django-DefectDojo/blob/unit-tests/unittests/test_apply_finding_template_placeholder.py) ? I'm kinda not used to Django testing we didn't pushed unittest cause I am struggling on our test assertion 😅

First, I highly recommend checking https://github.com/DefectDojo/django-DefectDojo/blob/master/readme-docs/DOCKER.md#running-the-unit-tests (just how to start tests - you do not need to run all of them, only yours `./dc-unittest.sh --profile ... --test-case unittests.test_apply_finding_template_placeholder`)

I see, you found inspiration in `test_apply_finding_template.py` but this specific one is not the best (might be written much better - simplified and access via API for basic tests is not necessary).

Find inspiration in 
- `test_cleanup_alerts.py` - for use-case "what will happen when I call `apply_cwe_mitigation`"
- `test_api_sonarqube_updater.py`- for use-case "if I call `template_replace_placeholder` with specific parameters, what will be the result?"

**Comment by mtesauro** [comment]
Closed as stale

**Pull request closed without merging**